src/main/java/com/bank/moneytransfer/datastore/BankAccountStorage.java

```diff
@@ -35,7 +37,19 @@ public BankAccount getBankAccount(Integer id) {
         return bankAccounts.values();
     }
 
-    public void updateBankAccount(Integer id, BigDecimal balance) {
-        bankAccounts.put(id, new BankAccount(id, balance));
+    //must be called in a synchronized block to ensure data integrity
+    public void updateBankAccounts(Integer fromId, Integer toId, BigDecimal amount)
+            throws FundsInsufficientTransferException {
+
+        //here it's guaranteed that both the accounts exist
+        BankAccount from = getBankAccount(fromId);
+        //existing balance must be greater than the transfer amount
+        if (from.getBalance().compareTo(amount) < 0) {
+            throw new FundsInsufficientTransferException(ErrorMessages.FUNDS_INSUFFICIENT_TRANSFER.getValue());
+        }
+        from.setBalance(from.getBalance().subtract(amount));
+
+        BankAccount to = getBankAccount(toId);
+        to.setBalance(to.getBalance().add(amount));
```

Comment:
updateBankAccount() was creating a new object on every transferMoney call instead of **actually** updating the bankAccount object. This led to intermittent failures in parallel integration tests. The new updateBankAccounts() correctly updates the bankAccounts when transferMoney request is done.